x-pack/plugins/lens/public/editor_frame_service/editor_frame/workspace_panel/workspace_panel.tsx

```diff
@@ -487,16 +493,27 @@ export const VisualizationWrapper = ({
     [context]
   );
   const searchSessionId = useLensSelector(selectSearchSessionId);
-
+  const datasourceLayers = useLensSelector((state) => selectDatasourceLayers(state, datasourceMap));
   const dispatchLens = useLensDispatch();
 
   const onData$ = useCallback(
     (data: unknown, adapters?: Partial<DefaultInspectorAdapters>) => {
       if (adapters && adapters.tables) {
-        dispatchLens(onActiveDataChange({ ...adapters.tables.tables }));
+        const [defaultLayerId] = Object.keys(datasourceLayers);
+
+        dispatchLens(
+          onActiveDataChange(
+            Object.entries(adapters.tables?.tables).reduce<Record<string, Datatable>>(
+              (acc, [key, value], index, tables) => ({
+                [tables.length === 1 ? defaultLayerId : key]: value,
+              }),
+              {}
+            )
+          )
+        );
       }
     },
-    [dispatchLens]
+    [datasourceLayers, dispatchLens]
```

Comment:
This will cause a lot of re-evaluations of this callback as `datasourceLayers` changes often. Let's move the default layer id calculation (`const [defaultLayerId] = Object.keys(datasourceLayers);`) into the component itself and pass `defaultLayerId` into the hook as a dependency instead.